test/testAdaptors/CompoundV3.t.sol

```diff
@@ -0,0 +1,974 @@
+
+        // Large whale deposits WETH to borrow remaining USDC.
+        uint256 whaleWeth = priceRouter.getValue(USDC, 2 * usdcInComet, WETH);
+        deal(address(WETH), address(this), whaleWeth);
+        WETH.approve(address(usdcComet), whaleWeth);
+        usdcComet.supply(address(WETH), whaleWeth);
+
+        // Have whale borrow all USDC from comet.
+        usdcComet.withdraw(address(USDC), usdcInComet);
+
+        assertApproxEqAbs(
+            cellar.totalAssetsWithdrawable(),
+            initialAssets,
+            2,
+            "Only initial assets should be withdrawable."
+        );
+
+        // Whale repays 10% of assets.
+        USDC.approve(address(usdcComet), assets / 10);
+        usdcComet.supply(address(USDC), assets / 10);
+
+        uint256 expectedAssets = initialAssets + assets / 10;
+
+        assertApproxEqAbs(
+            cellar.totalAssetsWithdrawable(),
+            expectedAssets,
+            2,
+            "Initial assets and 10% of assets should be withdrawable."
+        );
+
+        // Strategist withdraws as much as possible.
+        Cellar.AdaptorCall[] memory data = new Cellar.AdaptorCall[](1);
+        {
+            bytes[] memory adaptorCalls = new bytes[](1);
+            adaptorCalls[0] = _createBytesDataToWithdrawBaseFromCompoundV3(usdcComet, type(uint256).max);
+            data[0] = Cellar.AdaptorCall({ adaptor: address(supplyAdaptor), callData: adaptorCalls });
+        }
+        cellar.callOnAdaptor(data);
+
+        uint256 usdcInCellar = USDC.balanceOf(address(cellar));
+        assertApproxEqAbs(usdcInCellar, expectedAssets, 1, "Cellar should hold expected assets of USDC.");
+    }
+
+    function testIsLiquid(uint256 assets) external {
+        assets = bound(assets, 0.1e6, 1_000_000e6);
+        deal(address(USDC), address(this), assets);
+        bool isLiquid = false;
+        // Add cUSDCV3 Supply position, and set as holding position.
+        cellar.addPosition(0, usdcCompoundV3SupplyPosition, abi.encode(isLiquid), false);
+        cellar.setHoldingPosition(usdcCompoundV3SupplyPosition);
+
+        // Deposit into Cellar.
+        cellar.deposit(assets, address(this));
+
+        assertApproxEqAbs(
+            cellar.totalAssetsWithdrawable(),
+            initialAssets,
+            1,
+            "Assets withdrawable should equal initialAssets."
+        );
+
+        vm.startPrank(address(cellar));
+        bytes memory callData = abi.encodeWithSelector(
+            SupplyAdaptor.withdraw.selector,
+            0,
+            address(1),
+            abi.encode(0),
+            abi.encode(isLiquid)
+        );
+        vm.expectRevert(bytes(abi.encodeWithSelector(BaseAdaptor.BaseAdaptor__UserWithdrawsNotAllowed.selector)));
+        address(supplyAdaptor).functionDelegateCall(callData);
+        vm.stopPrank();
+    }
+
+    function testRecoveringFromTooManyCollateralsAddedToComet(uint256 assets) external {
```

Comment:
Can you elaborate on the purpose of this test?

Reply:
So the V3 helper logic will immediately return 0 if the number of assets in a comet is larger than an immutable value set on adaptor deployment. This is to protect the cellar from malicious Compound V3 governance that maxes out the number of assets in a comet, in an effort to expend more gas in the for loop in `getAccountHealthFactor`. If this does happen, the Cellar can be recovered by repaying all debt owed.

Reply:
Note Compound V3 does have a max collateral count logic, but I found it easier to verify gas safety by adding a check to our adaptor.

Reply:
Great, thank you for the explanation. This makes more sense now.